src/Microsoft.ML.Sweeper/AsyncSweeper.cs

```diff
@@ -208,7 +209,7 @@ public DeterministicSweeperAsync(IHostEnvironment env, Options options)
             _lock = new object();
             _results = new List<IRunResult>();
             _nullRuns = new HashSet<int>();
-            _paramQueue = new BufferBlock<ParameterSetWithId>();
+            _paramQueue = Channel.CreateUnbounded<ParameterSetWithId>();
```

Comment:
Are there any options we can set here? Specifically I think we can set [https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channeloptions.singlewriter?view=netcore-3.1](https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channeloptions.singlewriter?view=netcore-3.1)

Reply:
Do you think there would be a single reader as well, or just a single writer?

Reply:
I haven't extensively looked through the code, but from my initial reading I don't think we could guarantee a single reader. Two threads could call `ProposeAsync()` at the same time.

The single writer is guaranteed by the `lock(_lock)` in the `public void Update(int id, IRunResult result)` method. (And the fact that the only other place `PrepareNextBatch` is called is from inside the constructor - which can't have multiple threads.)